### ISSUE TITLE
Functional Backward Pass for Parallel Training

### ISSUE THREAD

**Issue by bwasti** [facebookresearch/shumai#47]
Right now `backward()` does two things:

1. Attaches a `grad` to all Tensors with `requires_grad === true`
2. Returns a list of Tensors that have a newly attached `grad`

Optimizers can then use this list of Tensors to update parameters accordingly.  Then, they update `grad` back to `null` to indicate they've finished updating the parameters.

However, this design is completely broken when done in parallel.  Two optimizers may receive the same list and conflict with each other during the update process.  Instead, we should return a list of Tensors *and* gradients.  This way, optimizers can update tensors in place (the one and **only** non-functional operation in the entirety of Shumai "`update`" will need to hold a lock) without worrying about conflict.

**Comment by bwasti** [comment]
fixed: https://github.com/facebookresearch/shumai/commit/f74cd9d6a6d6514c001b6ac897be5374b0bd4cb0
